### ISSUE TITLE
`stack setup` does not resume failed download.

### ISSUE THREAD

**Issue by weakish** [commercialhaskell/stack#2899]
If `ghc` is not installed, `stack setup` will download `ghc`.
If the download interrupted, run `stack setup` will download from the beginning,
instead of resuming unfinished download.

**Comment by mgsloan** [comment]
Makes sense.  I think the priority is that this works correctly - adding incrementality increases complexity.  So, not sure if it's worthwhile to make this change.

What do other folks think?

**Comment by lwm** [comment]
> So, not sure if it's worthwhile to make this change.

:+1:

**Comment by mgsloan** [comment]
Gonna go ahead and close this, then!

**Comment by baltar** [comment]
Can this be re-opened? I think it's fair to say that resuming downloads is a widely available and expected technology in 2018. 

To add a personal perspective, I'm on an unstable internet connection and trying to `stack setup` unsuccessfully for 3 days now.

**Comment by ekosuhariyadi** [comment]
Yes, this is kinda frustrating.
My connection unreliable, make the download stuck on 50% something and not continuable when connection available again, so I have to restart download ever and ever

**Comment by joecrayne** [comment]
Yes, please re-open.

+1

**Comment by afcady** [comment]
Well I'm currently waiting for a re-download of the first 97% of ghc so I decided to file a bug report here.

Let me say it's a very bad _look_ to have a downloader that does not resume.  My system already has multiple implementations of HTTP download that support resume, including `wget`, `curl`, `libcurl`, and many others.  Instead of calling out to one of the many existing robust implementations, an inferior self-rolled version is used by `stack`.

If you're going to roll your own, you have to make it at least good enough that it doesn't visibly waste the user's time.

Although the data in this case is far from precious, this is structurally a data loss bug.  You're deleting the partial download -- for no reason.  Just use a superior implementation.  Literally just call out to `curl` if it's available.  It could probably be implemented in the time it is taking me to wait for this second download.  Currently at 56% and it's slow as hell.

**Comment by mbj** [comment]
@afcady 

> t could probably be implemented in the time it is taking me to wait for this second download

do so, send a PR and get it accepted. Also you can apply for a refund on the money you spend on `stack`.

**Comment by mpilgrem** [comment]
I think this proposed change is beyond my own capability and I agree with @mgsloan that the priority is something that works correctly. If it helps others:

* Currently, Stack uses `Network.HTTP.Download.verifiedDownload` from the `http-download` package (see `Network.HTTP.StackClient.verifiedDownload`). It does so using `RetryPolicy` `Network.HTTP.Download.drRetryPolicyDefault` which is 'retry seven times with exponential backoff starting from one hundred milliseconds'. I appreciate that is about 'retrying', not 'continuing'.

* In respect of cross-platform considerations, the current Stack-supplied MSYS2 includes `wget.exe` (version 1.21.3) and `curl.exe` (version 7.84.0).

**Comment by afcady** [comment]
"Working correctly" means it only loses data that can be replaced?

It's failing when it does not have to because of a lack of robustness.  Calling out to `wget -c` would fix the issue.

> do so, send a PR and **get it accepted**

LOL, one of those things is not like the others ;)

I can patch the code to call out to wget (falling back to what's already implemented if wget is unavailable).  It's up to someone else to decide whether that would go into the code base.

**Comment by mpilgrem** [comment]
I consulted a GHCup (the Haskell installer) developer, to understand what GHCup currently does. It (a) uses `wget`/`curl`; but (b) does not continue interrupted downloads.

**Comment by hasufell** [comment]
I can implement resumable downloads in ghcup as a PoC. I think it's an interesting feature.

The challenge is how to figure out when to resume... we need to distinguish interrupt by signals, fatal download failure etc. I'm not sure if curl handles that well enough.

**Comment by weakish** [comment]
> The challenge is how to figure out when to resume... we need to distinguish interrupt by signals, fatal download failure etc.

I think a simpler approach is always resume, unless the user explicitly specifying to start over, for example, via a `--clean-cache` flag.
 
After all, corner cases, such as a middle man between HTTPS transfer changed the transfer content (then later the middle man stops attacking) or a local program mistakenly changed the already downloaded part, are rare.
